### PR TITLE
feat: support usage of globs in cli mode

### PR DESCRIPTION
**Pull request by thepassle**
**EDIT:** I seemed to have fixed the issue 🙂 The PR can be reviewed now

image
🎉

~~this PR isnt quite finished, it seems to be almost working, but I'm running into one issue that I cant quite figure out, I was hoping you could give me some pointers on that one, @devongovett ?~~

~~The issue is the following:~~

```
error[E0308]: `if` and `else` have incompatible types
   --> src/main.rs:252:11
    |
249 |           let css_modules_filename = if let Some(name) = css_modules {
    |  ____________________________________-
250 | |           name
    | |           ---- expected because of this
251 | |         } else {
252 | |           infer_css_modules_filename(&output_file)?
    | |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | |           |
    | |           expected `&std::string::String`, found struct `std::string::String`
    | |           help: consider borrowing here: `&infer_css_modules_filename(&output_file)?`
253 | |         };
    | |_________- `if` and `else` have incompatible types

For more information about this error, try `rustc --explain E0308`.
error: could not compile `lightningcss` due to previous error
```

fixes https://github.com/parcel-bundler/lightningcss/issues/90

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -66,26 +67,41 @@ struct SourceMapJson<'a> {
...
-      (filename, contents)
-    }
+  match &cli_args.input_file {
+    Some(g) => {
+      let globs = glob(g);
+
+      match glob(g) {
+        Ok(file_paths) => {
+          for file in file_paths {
+            if let Ok(f) = file {
```
Comment (thepassle) [inline src/main.rs:89]: I was also wondering if you could give me some pointers on this code, there's quite a bit of nested pattern matching going on, is there a cleaner/more idiomatic way to approach this?

File: src/main.rs [hunk 2]
```diff
@@ -94,157 +110,164 @@ pub fn main() -> Result<(), std::io::Error> {
...
+  }
 
-  let css_modules = if let Some(_) = cli_args.css_modules {
-    let pattern = if let Some(pattern) = cli_args.css_modules_pattern.as_ref() {
-      match lightningcss::css_modules::Pattern::parse(pattern) {
-        Ok(p) => p,
-        Err(e) => {
-          eprintln!("{}", e);
-          std::process::exit(1);
+  println!("{:#?}", &files);
```
Comment (thepassle) [inline src/main.rs:117]: I'll clean these up later, I was using them to debug while developing locally 🙃

File: src/main.rs (hunk 1 above)
Comment (devongovett) [reply src/main.rs:86]: One question I have is if we need to expand globs ourselves here or not? Usually the shell (e.g. bash) will expand globs before executing the program, so the glob itself will never reach it, it'll just be a list of files.

Comment (thepassle) [reply src/main.rs:86]: During development I used the following command to run the cli:
```
⚡ cargo run --features="cli" "*.css"
    Finished dev [unoptimized + debuginfo] target(s) in 0.58s
     Running `target/debug/lightningcss '*.css'`
html {
  color: red;
}

body {
  background: green;
}
```

Having the quotes around the glob does seem needed, if you provide the glob without quotes, it does indeed get expanded by the shell, and result in the following:

```
⚡ cargo run --features="cli" *.css
    Finished dev [unoptimized + debuginfo] target(s) in 0.42s
     Running `target/debug/lightningcss test.css test2.css`
error: Found argument 'test2.css' which wasn't expected, or isn't valid in this context

USAGE:
    lightningcss [OPTIONS] [INPUT_FILE]

For more information try --help
```

**Comment by devongovett** [comment]
Hey sorry it took so long for me to get back to you on this. I think we will need to change a few things.

At the moment, this will take only glob literal, which you need to quote to prevent your shell from expanding them. It would be nice if you could do `lightningcss a.css b.css` too. Then we wouldn't need to implement globs ourselves either, because `lightningcss *.css` would be expanded by your shell into multiple files and we get glob support for free.

In addition, at the moment this will compile each file, but then output them all concatenated together to stdout. If you use the `-o` option to output to a file, they will overwrite each other so only the last one is output. We'll need a way to somehow output each file with a different filename, perhaps by specifying a directory and preserving the filename of each input file.

**Pull request closed without merging**